### DISCUSSION TITLE
How to sync 2 ScrollViewers along the x axis inside a DataTemplate?

### DISCUSSION THREAD

**Discussion by donnyv** [AvaloniaUI/Avalonia#8046] [Q&A] [answered]
I'm making a custom table control. It has 2 ScrollViewers, one for the rows section and the other hidden in the columns section. They are both sitting inside a DataTemplate. I would like to scroll the columns section horizontal when the user scrolls the rows section horizontal. I tried Binding the Offset property to one another but couldn't figure out how to do it only to the X axis. Is this possible on the xaml side or does it have to happen in C#? Either way is fine, just trying to figure out how to do it.

**Comment by timunie** [comment, marked as answer]
Hi @donnyv 

just some thoughts from my side (I can't promise that any of them will work for you): 
- Have a common ViewModel with a property for the offset (haven't tested it). This property needs to bind twoway for one ScrollViewer and oneway for the other.
- Using a Behavior for this https://github.com/wieslawsoltes/AvaloniaBehaviors
- Create a new Control for this instead of a DataTemplate which handles it. This may have other benefits for you as well https://docs.avaloniaui.net/docs/authoring-controls/types-of-control#templated-controls
- Test the new TreeDataGrid https://docs.avaloniaui.net/docs/controls/treedatagrid
- Try out this control: https://github.com/wieslawsoltes/DataBox 

Happy coding
Tim

**Reply by donnyv** [reply]
Yes I'm making this inside a ControlTemplate. I'm assuming there must be a simple way to bind the 2 ScrollViewer controls using the Offset property. I just don't know what the syntax is. I will be adding more custom features specific to the app I'm making then the TreeDataGrid or DataBox offer.
```
<ResourceDictionary xmlns="https://github.com/avaloniaui"
                    xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml"
					xmlns:viewmodels="clr-namespace:DA.Studio.Controls.JsonTableControl"
					xmlns:sys="clr-namespace:System;assembly=mscorlib">

	<!--<viewmodels:TableViewModel x:Key="TableViewModel"/>-->

	<DataTemplate x:Key="TableViewModelDataTemplate">
		<Grid RowDefinitions="30,*">
			<ScrollViewer Name="HeaderScroller" Grid.Row="0" HorizontalScrollBarVisibility="Auto">
				<ItemsRepeater Items="{Binding Path=Header.Headers}">
					<ItemsRepeater.Layout>
						<StackLayout Spacing="0" Orientation="Horizontal" />
					</ItemsRepeater.Layout>
					<ItemsRepeater.ItemTemplate>
						<DataTemplate>
							<Button Content="{Binding .}" Background="LightGray" Foreground="White" Width="80" VerticalAlignment="Center" HorizontalAlignment="Center"></Button>
						</DataTemplate>
					</ItemsRepeater.ItemTemplate>
				</ItemsRepeater>
			</ScrollViewer>
			<ScrollViewer Name="RowScroller" HorizontalScrollBarVisibility="Auto" VerticalScrollBarVisibility="Auto" Grid.Row="1">
				<!--ScrollChanged="RowScrollChangedHandler"-->

				<ScrollViewer.Styles>
					<Style Selector="ScrollBar:horizontal">
						<Setter Property="Height" Value="15"/>
					</Style>
					<Style Selector="ScrollBar:vertical">
						<Setter Property="Width" Value="15"/>
					</Style>
				</ScrollViewer.Styles>
				<ItemsRepeater Items="{Binding Path=Rows}">
					<ItemsRepeater.Layout>
						<StackLayout Spacing="0" Orientation="Vertical" />
					</ItemsRepeater.Layout>
					<ItemsRepeater.ItemTemplate>
						<DataTemplate>
							<ItemsRepeater Items="{Binding Path=Cells}">
								<ItemsRepeater.Layout>
									<StackLayout Spacing="0" Orientation="Horizontal" />
								</ItemsRepeater.Layout>
								<ItemsRepeater.ItemTemplate>
									<DataTemplate>
										<Border Width="80" Height="40" BorderThickness="1,1,1,0" BorderBrush="Gray">
											<TextBlock Text="{Binding .}" VerticalAlignment="Center" HorizontalAlignment="Left" Margin="5"/>
										</Border>
									</DataTemplate>
								</ItemsRepeater.ItemTemplate>
							</ItemsRepeater>
						</DataTemplate>
					</ItemsRepeater.ItemTemplate>
				</ItemsRepeater>
			</ScrollViewer>
		</Grid>
	</DataTemplate>
	
	<ControlTemplate x:Key="JsonTableControlTemplate" TargetType="viewmodels:JsonTableControl">
		<ContentPresenter Name="ContentTablePresenter"></ContentPresenter>
	</ControlTemplate>
	
</ResourceDictionary>
```

**Reply by timunie** [reply]
Hi  @donnyv , 

This works for me: 
```xml
<StackPanel Spacing="5">
  <ScrollViewer x:Name="ScrollViewer1" HorizontalScrollBarVisibility="Visible" Height="50">
    <Border MinWidth = "2000" Background="{DynamicResource SystemAccentColor}">
    <TextBlock>This is a very very long Text in ScrollViewer 1. This is a very very long Text in ScrollViewer 1. This is a very very long Text in ScrollViewer 1. This is a very very long Text in ScrollViewer 1.  </TextBlock>
    </Border>
  </ScrollViewer>
  <ScrollViewer x:Name="ScrollViewer2" HorizontalScrollBarVisibility="Visible" Height="50"
                Offset="{Binding #ScrollViewer1.Offset, Mode=TwoWay}">
    <Border MinWidth = "2000" Background="{DynamicResource SystemAccentColorDark2}">
      <TextBlock>This is a very very long Text in ScrollViewer 2. This is a very very long Text in ScrollViewer 2. This is a very very long Text in ScrollViewer 2. This is a very very long Text in ScrollViewer 2.  </TextBlock>
    </Border>
  </ScrollViewer>
</StackPanel>
```

Result: 
![image](https://user-images.githubusercontent.com/47110241/165538102-be83941e-a012-4081-9a53-1503e85a7546.png)

Happy coding
Tim

**Reply by donnyv** [reply]
That worked! I was over thinking it. I thought I needed to map the Offset.x property. Didn't even think of mapping the whole object.

**Comment by RusselRillema** [comment]
HI @donnyv 
We're actually also looking to write a custom data grid control and we were struggling with a similar thing. Binding to the whole Offset definitely helped. Currently we are running into an issue of the bottom block also having a vertical scroll bar and thus the horizontal scroll bars don't line up at the end. I think we can solve this with some padding though. 

Anyway, the reason for the comment is that I wanted to reach out and ask if you managed to get a complete implementation for the grid and to ask if you'd be open to us talking through your approach and hearing about any "got ya's" you may have experienced while doing this.

We're migrating from a windows forms application where we'd built an extremely versatile grid with complex grouping, aggregation column customization etc. and I think it would be really helpful to ask you a couple questions.

If you'd be open to having a discussion and seeing if there's any overlap between what we're trying to achieve please reach out to me on 	russel@sandwichtrading.com.

It'd be awesome to hear from you.
Cheers
Russel
